Rails 5-1

jrafanie opened the pull request:
### Remaining items

- [x] Extract commits that can land on master for both rails 5.0 and 5.1 (this repo and others)
- [x] Fix julian to gregorian time issue
- [x] Fix deadlock/database lock test
- [x] Figure out what to do with rails 5-1-stable branch (5.1.7 was released)
- [x] Deploy to an appliance and look for failures/warnings
- [x] Verify debug and compile assets (enabled compile in test / monkey patched sprockets-rails)
- [x] Verify haml/haml_lint/erubi change (extracted to https://github.com/ManageIQ/manageiq/pull/18868)
- [x] Write up talk article with the most common rails 5.1 issues [DONE, here](http://talk.manageiq.org/t/riding-the-rails-5-1-train/4336)

Other repos:
- [x] Test UI classic repo with this fork as the manageiq repo

jrafanie commented on spec/models/vmdb_database_connection_spec.rb:
```diff
@@ -1,16 +1,12 @@
-    @db = FactoryGirl.create(:vmdb_database)
-  end
-
+  self.use_transactional_tests = false
```
Fixed based on @NickLaMuro's suggestion.  This is MUCH cleaner.  🙇 👏

jrafanie commented on app/models/chargeback.rb:
```diff
@@ -136,11 +136,14 @@ def new_chargeback_calculate_costs(consumption, rates)
 
+      # TODO: duration_of_report_step is 30.days for price plans but for consumption history,
+      # it's used for date ranges and needs to be 1.month with rails 5.1
+      duration = @options.interval == "monthly" ? 30.days : @options.duration_of_report_step
```
😖 🤢

kbrock commented:
ancestry relation updates in #18164

Fryguy commented on spec/models/chargeback_vm_spec.rb:
```diff
@@ -1398,3 +1398,30 @@ def find_result_by_vm_name_and_region(chargeback_result, vm_name, region)
 end
+
+describe Chargeback::Consumption do
+  context "#hours_in_month" do
```
Can you add a

```ruby
# TODO: Move to manageiq-consumption repo
```

kbrock commented:
Regarding this commit:

>update MiqReport::Search to include all columns

Virtual attributes were extracted from manageiq proper so that we could better nail down the requirements to get virtual attributes working in future versions of the product.

in a perfect world, I would like us upgrading rails 5.1 without requiring any virtual attribute hacks.
Our reporting pushes virtual attributes the furthest out of any of our components. Probably with rbac a close second. So while upgrading to rails 5.1 and 5.2, expect those to speak out and show where we need a little extra love

jrafanie commented:
Related: 
* replaced erubis with erubi for winrm:  https://github.com/WinRb/WinRM/pull/300
* replaced changed? to use will save changes/saved changes to in ancestry: https://github.com/stefankroes/ancestry/pull/434

kbrock commented on Gemfile:
```diff
@@ -20,15 +20,16 @@ def manageiq_plugin(plugin_name)
 # Unmodified gems
-gem "activerecord-virtual_attributes", "~>1.0.0"
-gem "activerecord-id_regions",        "~>0.2.0"
+gem "activerecord-virtual_attributes", :github => "ManageIQ/activerecord-virtual_attributes"
```
sigh

jrafanie replied:
just until it's released

jrafanie commented on config/environments/test.rb:
```diff
@@ -22,7 +22,8 @@
 
-  config.assets.compile = ENV['TEST_SUITE'] == 'spec:javascript'
+  config.assets.debug = true
+  config.assets.compile = %w(spec:javascript spec).include?(ENV['TEST_SUITE'])
```
Will need UI review on keeping these two changes.  We need to compile assets so tests don't raise a warning on missing assets from the pipeline.

jrafanie replied:
cc @h-kataria @himdel

jrafanie replied:
I noticed that the disabling of live compilation in other test suites causes a failure like:
```
DEPRECATION WARNING: The asset "layout/brand.svg" is not present in the asset pipeline.Falling back to an asset that may be in the public folder.
This behavior is deprecated and will be removed.
To bypass the asset pipeline and preserve this behavior,
use the `skip_pipeline: true` option.
 (called from image_path at /home/travis/build/jrafanie/manageiq-api/app/controllers/api/api_controller.rb:89)
```

Seen in https://github.com/ManageIQ/manageiq-api/pull/606

jrafanie commented on lib/extensions/ar_migration.rb:
```diff
@@ -1,7 +1,6 @@
   def self.discover_schema_migrations_ran_class
-    raise "Remove 'silence' block now that Rails version is >= 5.1" if Rails.version >= "5.1"
-    ActiveSupport::Deprecation.silence { return unless ActiveRecord::Base.connection.table_exists?("schema_migrations_ran") }
+    return unless ActiveRecord::Base.connection.table_exists?("schema_migrations_ran")
```
This revert will be removed before merge... it allowed us to see these deprecations in travis

jrafanie commented on Gemfile:
```diff
@@ -230,7 +232,7 @@ end
   group :development do
     gem "foreman"
-    gem "haml_lint",           "~>0.20.0", :require => false
+    gem "haml_lint",           "~>0.22.0", :require => false
```
@himdel @h-kataria I'll need you to review the haml/haml_lint change needed for erubi.  Please see the relevant commit.

jrafanie commented on config/environments/production.rb:
```diff
@@ -61,7 +61,7 @@
 
   # Send deprecation notices to registered listeners
-  config.active_support.deprecation = :notify
+  config.active_support.deprecation = :log
```
This will be dropped before merge

jrafanie commented:
@himdel @h-kataria @Fryguy @bdunne this is ready to go.  The talk article was written.  Is there anything else we need?

Here's the [talk article](http://talk.manageiq.org/t/riding-the-rails-5-1-train/4336).

These can be merged with rails 5.0 (now) as they will go red when 5.1 lands:

- [ ] https://github.com/ManageIQ/manageiq-providers-vmware/pull/407
- [ ] https://github.com/ManageIQ/manageiq-providers-ovirt/pull/377

The only other remaining items can only be merged after 5.1 is merged so there's no 5.0/5.1 compatible way to fix these:

- [ ] https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/174
- [ ] https://github.com/ManageIQ/manageiq-automation_engine/pull/327
- [ ] https://github.com/ManageIQ/manageiq-providers-openstack/pull/477

himdel commented:
Oh, correction, specs are passing locally.
https://github.com/ManageIQ/manageiq-ui-classic/pull/5676 is actually red now, restarting

himdel commented:
Oh, @jrafanie can you rebase please?
I think the only reason https://github.com/ManageIQ/manageiq-ui-classic/pull/5676 is failing travis is that this branch is based on master before the UI travis fix in https://github.com/ManageIQ/manageiq/pull/18892

jrafanie commented:
> Oh, @jrafanie can you rebase please?
> I think the only reason [ManageIQ/manageiq-ui-classic#5676](https://github.com/ManageIQ/manageiq-ui-classic/pull/5676) is failing travis is that this branch is based on master before the UI travis fix in #18892

thanks, I thought I did before court this morning but I rebased the manageiq rails-5-1 branch... ETOOMANYREPOS... I just rebased ui-classi's rails-5-1 test branch... will post here when it is green

jrafanie commented:
oops, comment is too close to close and comment

jrafanie commented:
ok, the [test rails 5.1 PR on ui classic](https://github.com/ManageIQ/manageiq-ui-classic/pull/5676) is now green after rebasing and force pushing this PR.
